### PR TITLE
Add custom @2x handling for mapbox v4 tile API when tileSize: 512

### PR DESCRIPTION
**Pull request by yhahn**
Native port of https://github.com/mapbox/mapbox-gl-js/pull/2044.

cc @jfirebaugh @kkaefer @brunoabinader

### CODE REVIEW

File: src/mbgl/util/mapbox.cpp [hunk 1]
```diff
@@ -135,7 +135,19 @@ std::string normalizeRasterTileURL(const std::string& url) {
...
-    normalizedURL.insert(extensionIdx, "{ratio}");
+
+    // Mapbox raster sources always use the @2x suffix on the v4 tile API
+    // to ensure a maximum 512 image size.
+    std::string::size_type versionIdx = url.rfind("/v4/");
+    if (tileSize == 512 &&
+        versionIdx != std::string::npos &&
+        (normalizedURL.compare(extensionIdx + 1, 3, "png") == 0 ||
+        normalizedURL.compare(extensionIdx + 1, 3, "jpg") == 0 ||
+        normalizedURL.compare(extensionIdx + 1, 4, "webp") == 0)) {
```
Comment (kkaefer) [inline src/mbgl/util/mapbox.cpp]: A better comparison would be add this function to `string.hpp`:

```cpp
inline bool beginsWith(const std::string& string, const std::string &prefix) {
    return std::equal(prefix.begin(), prefix.end(), string.begin());
}
```

**Comment by jfirebaugh** [comment]
Can we hold this until after #3715 lands? There's going to be significant conflicts.

**Comment by brunoabinader** [comment]
> Can we hold this until after #3715 lands?

:+1:

**Comment by jfirebaugh** [comment]
Updated for today's new offline world. @yhahn @tmpsantos Look good?

**Comment by tmpsantos** [comment]
> Updated for today's new offline world. @yhahn @tmpsantos Look good?

:+1:

**Pull request merged**